tests/optimize/test_qnspsa.py

```diff
@@ -0,0 +1,415 @@
+        # different order in sampling perturbation directions.
+        new_params_res, qnode_res = opt.step_and_cost(qnode, *params)
+
+        # test the expectation value
+        qnode_expected = np.cos(params[0] / 2) ** 2 - np.sin(params[0] / 2) ** 2 * np.cos(params[1])
+        assert np.allclose(qnode_res, qnode_expected)
+
+        # test the next-step parameter
+        _, grad_dirs = target_opt._get_spsa_grad_tapes(qnode, params, {})
+        _, tensor_dirs = target_opt._get_tensor_tapes(qnode, params, {})
+        qnode_finite_diff = get_grad_finite_diff(params, finite_diff_step, grad_dirs).reshape(1, 1)
+        grad_expected = [
+            qnode_finite_diff / (2 * finite_diff_step) * grad_dir for grad_dir in grad_dirs
+        ]
+        # reshape the params list into a tensor to reuse the
+        # get_metric_from_single_input_qnode helper function
+        params_tensor = np.array(params).reshape(1, len(params))
+        metric_tensor_expected = get_metric_from_single_input_qnode(
+            params_tensor, finite_diff_step, tensor_dirs
+        )
+
+        # regularize raw metric tensor
+        identity = np.identity(metric_tensor_expected.shape[0])
+        avg_metric_tensor = 0.5 * (identity + metric_tensor_expected)
+        tensor_reg = np.real(sqrtm(np.matmul(avg_metric_tensor, avg_metric_tensor)))
+        tensor_reg = (tensor_reg + regularization * identity) / (1 + regularization)
+
+        inv_metric_tensor = np.linalg.inv(tensor_reg)
+        grad_tensor = np.array(grad_expected).reshape(
+            inv_metric_tensor.shape[0],
+        )
+        new_params_tensor_expected = params_tensor - stepsize * np.matmul(
+            inv_metric_tensor, grad_tensor
+        )
+
+        assert np.allclose(
+            np.array(new_params_res).reshape(new_params_tensor_expected.shape),
+            new_params_tensor_expected,
+        )
+
+    def test_step_and_cost_with_non_trainable_input(self, finite_diff_step, seed):
+        """Test step_and_cost() function with the qnode with non-trainable input."""
+        # TODO: The test largely duplicates the test of test_step_from_single_input. Futher refactoring
+        # might be possible.
+        regularization = 1e-3
+        stepsize = 1e-2
+        opt = qml.QNSPSAOptimizer(
+            stepsize=stepsize,
+            regularization=regularization,
+            finite_diff_step=finite_diff_step,
+            resamplings=1,
+            blocking=True,
+            history_length=5,
+            seed=seed,
+        )
+        # target opt is used to reproduce the result with step()
+        target_opt = deepcopy(opt)
+
+        qnode, params_shape = get_qnode_with_non_trainable_input()
+        params = np.random.rand(*params_shape)
+
+        new_params_res, qnode_res = opt.step_and_cost(qnode, params, "placeholder")
+        new_params_expected = target_opt.step(qnode, params, "placeholder")
```

Comment:
Can you use a non-trainable tensor instead of a string here?

Reply:
Changed to a tensor with ``requires_grad`` field turned off.